[Spring Data JPA] 오지현 미션 제출합니다.

zhy2on opened the pull request:
안녕하세요 리뷰어님! 이번 미션도 수고 많으셨습니다!
이번 미션은 저번보다 더 어려워서.... 많이 헤맸고 최적화 부분도 신경을 못 썼습니다....
저번 스터디 때 배운 내용으로 서브모듈로 설정파일을 관리했는데 리뷰를 위해 public으로 설정해두었습니다!
잘못된 부분 지적해주시면 감사하겠습니다ㅠㅠ...

zhy2on committed:
feat: 1단계

zhy2on committed:
feat: 2단계

zhy2on committed:
feat: 3단계

zhy2on committed:
feat: 설정파일 보호를 위한 서브모듈 추가

zhy2on committed:
feat: 4단계

zhy2on committed:
feat: 5단계

zhy2on committed:
feat: 6단계

mangsuyo commented on src/main/java/roomescape/member/MemberRepository.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.Optional;
+
+public interface MemberRepository extends JpaRepository<Member, Long> {
+    Optional<Member> findByEmailAndPassword(String email, String password);
+    Optional<Member> findByName(String name);
```
단건 조회 결과를 `Optional`을 활용하셨군요! 좋습니다 :)

스터디 공통 피드백과 같이 중복되는 예외처리를 `default`를 통해서 처리 해보시는 것도 좋을 것같습니다!

mangsuyo commented on src/main/java/roomescape/member/MemberRequest.java:
```diff
@@ -1,19 +1,4 @@
-        return password;
-    }
+public record MemberRequest(String name, String email, String password) {
 }
```
`record`를 이용해서 `DTO`를 생성해 getter와 같은 코드의 중복을 막으셨군요! 좋습니다 :)

mangsuyo commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -1,30 +1,102 @@
+            member = memberRepository.findByName(loginMember.name())
+                    .orElseThrow(() -> new RuntimeException("Member not found"));
+            name = member.getName();
+        }
```
저는 `loginMember`에 대해서는 예외처리 하지 않았는데 좋은 예외처리 배우고 갑니다!

mangsuyo commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+    List<Reservation> findByDateAndTheme_Id(String date, Long themeId);
+
+    @Query("SELECT r FROM Reservation r JOIN FETCH r.theme t JOIN FETCH r.time ti WHERE r.member.id = :memberId")
+    List<Reservation> findAllByMemberIdWithThemeAndTime(@Param("memberId") Long memberId);
```
`JPQL`를 사용하신 의도가 궁금합니다!

zhy2on replied:
``패치 조인``을 사용해서 N+1 문제를 해결해 보고자 했습니다..!
https://ugo04.tistory.com/11

여기선 해당 메서드를 이용해 response를 생성할 때, ``getTheme().getName()``, ``getTime().getValue()``를 사용하게 됩니다.
기존 지연 로딩 전략을 따르면 연관 데이터가 불러와지지 않은 상태이기 때문에 추가적인 쿼리를 날려서 Theme의 name과 Time의 value를 가져오게 됩니다.

 ``JOIN FETCH``를 사용해서 한 번의 쿼리에 모든 정보를 불러와서 추가적인 쿼리를 날리지 않는 것이 성능에 더 좋을 것이라 생각했습니다!

그런데 더 찾아보니 패치 조인을 하게 되면 전체 데이터가 불러와져서 Theme의 경우는 description 같은 불필요한 데이터까지 불러와지게 되는데 이 때 6단계에서 제공된 예시처럼
```java
@Query("SELECT new roomescape.reservation.ReservationResponse(r.id, r.name, t.name, r.date, ti.value) " +
       "FROM Reservation r JOIN r.theme t JOIN r.time ti WHERE r.member.id = :memberId")
List<ReservationResponse> findReservationsByMemberId(@Param("memberId") Long memberId);
```

이런식으로 ``DTO 프로젝션``을 사용해서 원하는 필드만 정확히 가져오는 게 더 나아보이네요.. 이 부분은 제가 공부를 더 해봐야할 것 같습니다ㅠㅠ..

혹시 다른 의견 있으시다면 댓글 남겨주시면 감사하겠습니다!

mangsuyo commented on src/main/java/roomescape/waiting/WaitingService.java:
```diff
@@ -0,0 +1,75 @@
+        Time time = timeRepository.findById(waitingRequest.time())
+                .orElseThrow(() -> new RuntimeException("Time not found"));
+
+        if (!isSlotReserved(waitingRequest.date(), theme.getId(), time.getId())) {
+            throw new IllegalStateException("해당 시간대에 예약이 없어 예약 대기를 할 수 없습니다.");
+        }
+
+        if (hasExistingWaiting(memberId, waitingRequest.date(), theme.getId(), time.getId())) {
+            throw new IllegalStateException("이미 예약 대기 중입니다.");
+        }
```
어플리케이션 흐름을 깊게 고민 하면서 코드를 작성하신 것 같아요 :)
예약 대기 중복을 막는 예외처리가 인상적입니다!

mangsuyo commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -1,30 +1,102 @@
+                        it.getDate(),
+                        it.getTime().getValue()
+                ))
                 .toList();
```
올바르게 줄바꿈이 되어있어 읽기에도 편한 것 같습니다 :)

mangsuyo commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -1,30 +1,102 @@
+    public List<MyReservationResponse> findAllByMemberId(Long memberId) {
+        List<MyReservationResponse> result = new ArrayList<>();
+
+        result.addAll(reservationRepository.findAllByMemberIdWithThemeAndTime(memberId).stream()
+                .map(it -> new MyReservationResponse(
+                        it.getId(),
+                        it.getTheme().getName(),
+                        it.getDate(),
+                        it.getTime().getValue(),
+                        "예약"
+                ))
+                .toList());
+
+        List<WaitingWithRank> waitings = waitingService.findWaitingsWithRankByMemberId(memberId);
+        result.addAll(waitings.stream()
+                .map(it -> new MyReservationResponse(
+                        it.getWaiting().getId(),
+                        it.getWaiting().getTheme().getName(),
+                        it.getWaiting().getDate(),
+                        it.getWaiting().getTime().getValue(),
+                        (it.getRank() + 1) + "번째 예약대기"
+                ))
+                .toList());
+
+        return result;
```
`reservation`과 `waiting`을 `MyReservationResponse`로 변환후 하나의 리스트에 합쳐 반환하는 코드네요!

`MyReservationResponse`에 정적 팩토리 메서드 `of`를 이용하면 `MyReservationResponse`로 변환하는 로직 중복을 줄이고 의도를 분명하게 전달할 수 있을 것 같아요!

zhy2on replied:
주신 의견을 바탕으로 리팩토링 해보겠습니다!! 감사합니다

mangsuyo reviewed (commented):
안녕하세요 지현님!

코드 내용과 구조가 깔끔하여 코드를 읽고 이해하는 데에 시간이 얼마 걸리지 않았습니다.
또한 예외처리를 보고 어플리케이션의 흐름을 잘 이해하고 계신 것 같다는 생각이 들었어요!

좋은 코드 잘 보고 갑니다 :)

zhy2on committed:
refactor: 정적메서드 of 을 이용하여 리팩토링

zhy2on committed:
chore: 코드 정렬

zhy2on committed:
refactor: RuntimeException 대신 상황에 맞는 익셉션 사용

zhy2on committed:
refactor: 리팩토링
- WaitingWithRank record로 변경
- raw ResponseEntity 제거
- 생성자 주입 받는 필드 중 final로 구현되지 않았던 부분 수정

zhy2on committed:
feat: 7단계

zhy2on committed:
feat: 7단계
